File: DiaryNutrition/app/src/main/java/ediger/diarynutrition/diary/water/AddWaterDialog.java
package ediger.diarynutrition.diary.water;

import android.app.Dialog;
import android.app.TimePickerDialog;
import android.content.Context;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.google.android.material.textfield.TextInputEditText;

import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.DialogFragment;
import androidx.appcompat.app.AlertDialog;
import androidx.lifecycle.ViewModelProvider;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.TextView;
import android.widget.TimePicker;
import android.widget.Toast;

import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.InterstitialAd;

import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Locale;

import ediger.diarynutrition.Consts;
import ediger.diarynutrition.PreferenceHelper;
import ediger.diarynutrition.R;
import ediger.diarynutrition.data.source.entities.Water;
import ediger.diarynutrition.databinding.DialogAddWaterBinding;
import ediger.diarynutrition.diary.DiaryViewModel;
import ediger.diarynutrition.util.SnackbarUtils;

import static ediger.diarynutrition.PreferenceHelper.*;

public class AddWaterDialog extends DialogFragment implements View.OnClickListener, View.OnLongClickListener {

    public static final String ARG_DATE = "date";

    private DialogAddWaterBinding mBinding;
    private DiaryViewModel mViewModel;

    private boolean isAdsRemoved;

    private Calendar mCurrentDatetime = Calendar.getInstance();
    private SimpleDateFormat timeFormatter = new SimpleDateFormat("kk:mm", Locale.getDefault());

    private InterstitialAd mInterstitialAd;

    private TimePickerDialog.OnTimeSetListener t = new TimePickerDialog.OnTimeSetListener() {
        public void onTimeSet(TimePicker view, int hourOfDay, int minute) {
            mCurrentDatetime.set(mCurrentDatetime.get(Calendar.YEAR),
                    mCurrentDatetime.get(Calendar.MONTH),
                    mCurrentDatetime.get(Calendar.DAY_OF_MONTH),
                    hourOfDay, minute);
            mBinding.edTime.setText(timeFormatter.format(mCurrentDatetime.getTime()));
        }
    };

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mViewModel = new ViewModelProvider(this).get(DiaryViewModel.class);
        setStyle(DialogFragment.STYLE_NORMAL, R.style.FullscreenDialog);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        mBinding = DataBindingUtil.inflate(inflater, R.layout.dialog_add_water, container, false);

        isAdsRemoved = PreferenceHelper.getValue(Consts.PREF_ADS_REMOVED, Boolean.class, false);

        mInterstitialAd = new InterstitialAd(getActivity());
        mInterstitialAd.setAdUnitId(getString(R.string.banner_ad_inter_water_id));

        if (!isAdsRemoved) {
            mInterstitialAd.loadAd(new AdRequest.Builder().build());
        }

        mCurrentDatetime.setTimeInMillis(getArguments().getLong(ARG_DATE, 0));
        mCurrentDatetime.set(Calendar.HOUR_OF_DAY, Calendar.getInstance().get(Calendar.HOUR_OF_DAY));
        mCurrentDatetime.set(Calendar.MINUTE, Calendar.getInstance().get(Calendar.MINUTE));

        return mBinding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mBinding.edWater.requestFocus();
        mBinding.setOnClickListener(this);
        mBinding.setOnLongClickListener(this);

        mBinding.edTime.setText(timeFormatter.format(mCurrentDatetime.getTime()));
        mBinding.edTime.setOnClickListener(v -> {
            hideKeyboard();
            new TimePickerDialog(getActivity(), t, mCurrentDatetime.get(Calendar.HOUR_OF_DAY),
                    mCurrentDatetime.get(Calendar.MINUTE), true).show();
        });

        mBinding.toolbar.inflateMenu(R.menu.menu_dialog_water);
        mBinding.toolbar.setNavigationOnClickListener(v -> {
            hideKeyboard();
            dismiss();
        });
        mBinding.toolbar.setOnMenuItemClickListener(item -> {
            addWater();
            return true;
        });

        //Toggle keyboard at start
        InputMethodManager im = (InputMethodManager) getActivity()
                .getSystemService(Context.INPUT_METHOD_SERVICE);
        im.toggleSoftInput(InputMethodManager.SHOW_FORCED, 0);
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
    }

    @Override
    public void onStart() {
        super.onStart();
        Dialog dialog = getDialog();
        if (dialog != null) {
            int width = ViewGroup.LayoutParams.MATCH_PARENT;
            int height = ViewGroup.LayoutParams.MATCH_PARENT;
            dialog.getWindow().setLayout(width, height);
        }
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.card_water_1:
                mViewModel.addWater(new Water(Integer.parseInt(mBinding.txtWater1.getText().toString()),
                        mCurrentDatetime.getTimeInMillis()));
                hideKeyboard();

                if (!isAdsRemoved && mInterstitialAd.isLoaded()) mInterstitialAd.show();

                dismiss();
                break;
            case R.id.card_water_2:
                mViewModel.addWater(new Water(Integer.parseInt(mBinding.txtWater2.getText().toString()),
                        mCurrentDatetime.getTimeInMillis()));
                hideKeyboard();

                if (!isAdsRemoved && mInterstitialAd.isLoaded()) mInterstitialAd.show();

                dismiss();
                break;
            case R.id.card_water_3:
                mViewModel.addWater(new Water(Integer.parseInt(mBinding.txtWater3.getText().toString()),
                        mCurrentDatetime.getTimeInMillis()));
                hideKeyboard();

                if (!isAdsRemoved && mInterstitialAd.isLoaded()) mInterstitialAd.show();

                dismiss();
                break;
            case R.id.card_water_4:
                mViewModel.addWater(new Water(Integer.parseInt(mBinding.txtWater4.getText().toString()),
                        mCurrentDatetime.getTimeInMillis()));
                hideKeyboard();

                if (!isAdsRemoved && mInterstitialAd.isLoaded()) mInterstitialAd.show();

                dismiss();
                break;
        }
    }

    @Override
    public boolean onLongClick(View v) {
        switch (v.getId()) {
            case R.id.card_water_1:
                showDialog(0);
                break;
            case R.id.card_water_2:
                showDialog(1);
                break;
            case R.id.card_water_3:
                showDialog(2);
                break;
            case R.id.card_water_4:
                showDialog(3);
                break;
        }
        return false;
    }

    @Override
    public void onResume() {
        super.onResume();
        waterValuesInit();
    }

    private void addWater() {
        if (mBinding.edWater.getText().toString().matches("")
                || mBinding.edWater.getText().toString().startsWith("0")) {
            Toast.makeText(getActivity(), getResources().getString(R.string.message_dialog_water),
                    Toast.LENGTH_LONG).show();
        } else {
            mViewModel.addWater(new Water(Integer.parseInt(mBinding.edWater.getText().toString()),
                    mCurrentDatetime.getTimeInMillis()));
            hideKeyboard();

            if (!isAdsRemoved && mInterstitialAd.isLoaded()) mInterstitialAd.show();

            dismiss();
        }

    }

    private void hideKeyboard() {
        View view = getView().findFocus();
        if (view != null) {
            InputMethodManager imm = (InputMethodManager) getActivity().getSystemService(Context.INPUT_METHOD_SERVICE);
            imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
        }
    }

    private void waterValuesInit() {
        String[] consts = {
                Consts.PREF_WATER_SERVING_1,
                Consts.PREF_WATER_SERVING_2,
                Consts.PREF_WATER_SERVING_3,
                Consts.PREF_WATER_SERVING_4,
        };

        TextView[] txtViews = {mBinding.txtWater1, mBinding.txtWater2, mBinding.txtWater3, mBinding.txtWater4};

        for (int i = 0; i < consts.length; i++) {
            int value = PreferenceHelper.getValue(consts[i], Integer.class, 0);
            if (value != 0) {
                txtViews[i].setText(String.valueOf(value));
            }
        }
    }

    private void showDialog(int cardId) {
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity(), R.style.DialogStyle);
        View root = getActivity().getLayoutInflater().inflate(R.layout.dialog_water_serving, null);

        builder.setView(root);
        builder.setTitle(getString(R.string.dialog_title_water_serving));

        final TextInputEditText edWater = root.findViewById(R.id.ed_water_serving);
        edWater.requestFocus();

        builder.setPositiveButton(R.string.dialog_change, (dialog, which) -> {
            if (edWater.getText().toString().matches("") || edWater.getText().toString()
                    .startsWith("0")) {
                SnackbarUtils.showSnackbar(getView(), getString(R.string.message_dialog_water));
            } else {
                switch (cardId) {
                    case 0:
                        mBinding.txtWater1.setText(edWater.getText().toString());
                        PreferenceHelper.setValue(Consts.PREF_WATER_SERVING_1,
                                Integer.parseInt(edWater.getText().toString()));
                        break;
                    case 1:
                        mBinding.txtWater2.setText(edWater.getText().toString());
                        PreferenceHelper.setValue(Consts.PREF_WATER_SERVING_2,
                                Integer.parseInt(edWater.getText().toString()));
                        break;
                    case 2:
                        mBinding.txtWater3.setText(edWater.getText().toString());
                        PreferenceHelper.setValue(Consts.PREF_WATER_SERVING_3,
                                Integer.parseInt(edWater.getText().toString()));
                        break;
                    case 3:
                        mBinding.txtWater4.setText(edWater.getText().toString());
                        PreferenceHelper.setValue(Consts.PREF_WATER_SERVING_4,
                                Integer.parseInt(edWater.getText().toString()));
                        break;
                }
            }
        });

        builder.setNegativeButton(R.string.dialog_cancel, (dialog, which) -> dialog.dismiss());
        builder.create();
        builder.show();
    }

}

File: DiaryNutrition/app/src/main/java/ediger/diarynutrition/diary/water/AddWaterDialog.kt
package ediger.diarynutrition.diary.water

import android.app.TimePickerDialog
import android.app.TimePickerDialog.OnTimeSetListener
import android.content.DialogInterface
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.View.OnLongClickListener
import android.view.ViewGroup
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.viewModels
import com.google.android.material.textfield.TextInputEditText
import ediger.diarynutrition.*
import ediger.diarynutrition.databinding.DialogAddWaterBinding
import ediger.diarynutrition.util.SnackbarUtils
import ediger.diarynutrition.util.hideKeyboard
import ediger.diarynutrition.util.showKeyboard
import java.text.SimpleDateFormat
import java.util.*

class AddWaterDialog : DialogFragment(), View.OnClickListener, OnLongClickListener {

    private lateinit var binding: DialogAddWaterBinding

    private val viewModel: WaterViewModel by viewModels()

    private val timeFormatter = SimpleDateFormat("kk:mm", Locale.getDefault())

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setStyle(STYLE_NORMAL, R.style.FullscreenDialog)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View {
        binding = DialogAddWaterBinding.inflate(inflater, container, false)

        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        viewModel.addTime = Calendar.getInstance().apply {
            timeInMillis = arguments?.getLong(ARG_DATE, viewModel.addTime.timeInMillis) ?: viewModel.addTime.timeInMillis
            set(Calendar.HOUR_OF_DAY, viewModel.addTime[Calendar.HOUR_OF_DAY])
            set(Calendar.MINUTE, viewModel.addTime[Calendar.MINUTE])
        }

        waterValuesInit()

        with(binding) {
            onClickListener = this@AddWaterDialog
            onLongClickListener = this@AddWaterDialog
            lifecycleOwner = viewLifecycleOwner

            toolbar.inflateMenu(R.menu.menu_dialog_water)
            toolbar.setNavigationOnClickListener {
                hideKeyboard()
                dismiss()
            }
            toolbar.setOnMenuItemClickListener {
                addWater(edWater.text.toString())
                true
            }

            edTime.setText(timeFormatter.format(viewModel.addTime.time))
            edTime.setOnClickListener {
                hideKeyboard()
                TimePickerDialog(
                        requireContext(),
                        { _, hourOfDay, minute ->
                            viewModel.addTime[Calendar.HOUR_OF_DAY] = hourOfDay
                            viewModel.addTime[Calendar.MINUTE] = minute
                            binding.edTime.setText(timeFormatter.format(viewModel.addTime.time))
                        },
                        viewModel.addTime[Calendar.HOUR_OF_DAY],
                        viewModel.addTime[Calendar.MINUTE], true
                ).show()
            }
        }
    }

    override fun onStart() {
        super.onStart()
        dialog?.let {
            val width = ViewGroup.LayoutParams.MATCH_PARENT
            val height = ViewGroup.LayoutParams.MATCH_PARENT
            it.window?.setLayout(width, height)
        }
    }

    override fun onClick(v: View) = with(binding) {
        val amount = when (v.id) {
            R.id.card_water_1 -> txtWater1.text.toString()
            R.id.card_water_2 -> txtWater2.text.toString()
            R.id.card_water_3 -> txtWater3.text.toString()
            R.id.card_water_4 -> txtWater4.text.toString()
            else -> ""
        }
        addWater(amount)
    }

    override fun onLongClick(v: View): Boolean {
        when (v.id) {
            R.id.card_water_1 -> showDialog(0)
            R.id.card_water_2 -> showDialog(1)
            R.id.card_water_3 -> showDialog(2)
            R.id.card_water_4 -> showDialog(3)
        }
        return false
    }

    override fun onResume() {
        super.onResume()

        activity?.showKeyboard(binding.edWater)
    }

    private fun addWater(amount: String) {
        if (amount.isEmpty()
                || amount.startsWith("0")) {
            Toast.makeText(activity, resources.getString(R.string.message_dialog_water), Toast.LENGTH_LONG).show()
        } else {
            viewModel.addWater(amount.toInt())
            hideKeyboard()
            dismiss()
        }
    }

    private fun hideKeyboard() {
        activity?.hideKeyboard(binding.root.findFocus())
    }

    private fun waterValuesInit() {
        val servingKeys = arrayOf(
                PREF_WATER_SERVING_1,
                PREF_WATER_SERVING_2,
                PREF_WATER_SERVING_3,
                PREF_WATER_SERVING_4)
        val txtViews = arrayOf(binding.txtWater1, binding.txtWater2, binding.txtWater3, binding.txtWater4)
        for (i in servingKeys.indices) {
            val value = PreferenceHelper.getValue(servingKeys[i], Int::class.javaObjectType, 0)
            if (value != 0) {
                txtViews[i].text = value.toString()
            }
        }
    }

    private fun showDialog(cardId: Int) {
        AlertDialog.Builder(requireActivity(), R.style.DialogStyle).apply {
            val root = requireActivity().layoutInflater.inflate(R.layout.dialog_water_serving, null)
            setView(root)
            setTitle(getString(R.string.dialog_title_water_serving))

            val edWater: TextInputEditText = root.findViewById(R.id.ed_water_serving)
            edWater.requestFocus()

            setPositiveButton(R.string.dialog_change) { _, _ ->
                if (edWater.text.toString().isEmpty() || edWater.text.toString().startsWith("0")) {
                    SnackbarUtils.showSnackbar(view, getString(R.string.message_dialog_water))
                } else {
                    when (cardId) {
                        0 -> {
                            binding.txtWater1.text = edWater.text.toString()
                            PreferenceHelper.setValue(PREF_WATER_SERVING_1, edWater.text.toString().toInt())
                        }
                        1 -> {
                            binding.txtWater2.text = edWater.text.toString()
                            PreferenceHelper.setValue(PREF_WATER_SERVING_2, edWater.text.toString().toInt())
                        }
                        2 -> {
                            binding.txtWater3.text = edWater.text.toString()
                            PreferenceHelper.setValue(PREF_WATER_SERVING_3, edWater.text.toString().toInt())
                        }
                        3 -> {
                            binding.txtWater4.text = edWater.text.toString()
                            PreferenceHelper.setValue(PREF_WATER_SERVING_4, edWater.text.toString().toInt())
                        }
                    }
                }
            }

            setNegativeButton(R.string.dialog_cancel) { dialog: DialogInterface, _: Int -> dialog.dismiss() }
        }.create().show()
    }

    companion object {
        const val ARG_DATE = "date"
    }
}